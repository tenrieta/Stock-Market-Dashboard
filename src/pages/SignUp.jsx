import React from "react";

// Bootstrap
import Container from "react-bootstrap/Container";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";
import Card from "react-bootstrap/Card";
import Form from "react-bootstrap/Form";
import Button from "react-bootstrap/Button";
import Alert from "react-bootstrap/Alert";

// Firebase
import { createUserWithEmailAndPassword } from "firebase/auth";
import { auth } from "../firebase";

const SignUp = function SignUp() {
  // Local state for Form Data
  const [formData, setFormData] = React.useState({
    formLoading: false,
    formError: null,
    formValidated: false,
    email: "",
    newPassword1: "",
    newPassword2: "",
  });

  // Change handler which updates local state
  // In order to make it universal for all fields, we dynamically use [fieldName]: fieldValue
  const handleChange = (event) => {
    const fieldName = event.target.name;
    const fieldValue = event.target.value;
    setFormData((prevState) => ({ ...prevState, [fieldName]: fieldValue }));
  };

  // Submit handler
  const handleSubmit = (event) => {
    // First preventing default form behavior
    event.preventDefault();

    // Validation with Bootstrap
    setFormData((prevState) => ({ ...prevState, formValidated: true }));
    if (event.currentTarget.checkValidity() === false) {
      event.stopPropagation();
      return null;
    }

    // Verification of passwords
    if (formData.password1 !== formData.password2) {
      setFormData((prevState) => ({
        ...prevState,
        formError: "The passwords don't match.",
      }));
      event.stopPropagation();
      return null;
    }

    // Loading state prevents accidental double clicks on the submit button
    setFormData((prevState) => ({ ...prevState, formLoading: true }));

    // Sending data to Firebase
    // When Firebase approves, creates account, and logs the user in, user data is dispatched to Redux from App.jsx, where we have a listener onAuthStateChanged
    createUserWithEmailAndPassword(auth, formData.email, formData.newPassword1)
      // If there is an error, we show it in the form
      .catch((firebaseError) => {
        setFormData((prevState) => ({
          ...prevState,
          formLoading: false,
          formValidated: false,
          formError: `There was an error: ${firebaseError.code}.`,
        }));
        // We also set a timer to clean the error message and form validation
        setTimeout(() => {
          setFormData((prevState) => ({ ...prevState, formError: "" }));
        }, 3000);
      });
    return null;
  };
  return (
    <Container>
      <Row className="justify-content-md-center">
        <Col className="col-lg-6 col-md-10 mx-auto">
          <Card className="mb-4">
            <Card.Body>
              <h2 className="text-center mb-4">Hello! Please sign up.</h2>
              {formData.formError && (
                <Alert variant="danger">{formData.formError}</Alert>
              )}
              <Form
                noValidate
                validated={formData.formValidated}
                onSubmit={handleSubmit}
              >
                <Form.Group id="email" className="mb-3">
                  <Form.Label>Email</Form.Label>
                  <Form.Control
                    type="email"
                    required
                    value={formData.email}
                    onChange={handleChange}
                    name="email"
                  />
                </Form.Group>
                <Form.Group id="password" className="mb-3">
                  <Form.Label>Password</Form.Label>
                  <Form.Control
                    type="password"
                    value={formData.newPassword1}
                    onChange={handleChange}
                    name="newPassword1"
                  />
                </Form.Group>
                <Form.Group id="password-confirm" className="mb-3">
                  <Form.Label>Password Confirmation</Form.Label>
                  <Form.Control
                    type="password"
                    value={formData.newPassword2}
                    onChange={handleChange}
                    name="newPassword2"
                  />
                </Form.Group>
                <div className="text-center">
                  <Button
                    disabled={formData.formLoading}
                    variant="primary"
                    id="submit-button"
                    type="submit"
                  >
                    Sign Up
                  </Button>
                </div>
              </Form>
            </Card.Body>
          </Card>
        </Col>
      </Row>
    </Container>
  );
};
export default SignUp;
