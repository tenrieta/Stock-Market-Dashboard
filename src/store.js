/* eslint-disable no-underscore-dangle */
import { applyMiddleware, compose, createStore } from "redux";
import thunk from "redux-thunk";
import rootReducer from "./reducers";

const store = createStore(
  rootReducer,
  compose(
    applyMiddleware(thunk), // Thunk is necessary for async action creators
    window.__REDUX_DEVTOOLS_EXTENSION__
      ? window.__REDUX_DEVTOOLS_EXTENSION__()
      : (f) => f // This is to make the Redux for Chrome extension work in an incognito window with disabled Redux extension
  )
);

export default store;
