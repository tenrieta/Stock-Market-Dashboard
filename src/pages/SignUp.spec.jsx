import React from "react";
import renderer from "react-test-renderer";
import { Provider } from "react-redux";
import store from "../store";

import SignUp from "./SignUp";

describe("SignUp", () => {
  test("SignUp snapshot renders", () => {
    const component = renderer.create(
      <Provider store={store}>
        <SignUp />
      </Provider>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
