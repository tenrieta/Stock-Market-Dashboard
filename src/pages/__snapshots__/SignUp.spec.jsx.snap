// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUp SignUp snapshot renders 1`] = `
<div
  className="container"
>
  <div
    className="justify-content-md-center row"
  >
    <div
      className="col-lg-6 col-md-10 mx-auto col"
    >
      <div
        className="mb-4 card"
      >
        <div
          className="card-body"
        >
          <h2
            className="text-center mb-4"
          >
            Hello! Please sign up.
          </h2>
          <form
            className=""
            noValidate={true}
            onSubmit={[Function]}
          >
            <div
              className="mb-3"
              id="email"
            >
              <label
                className="form-label"
              >
                Email
              </label>
              <input
                className="form-control"
                name="email"
                onChange={[Function]}
                required={true}
                type="email"
                value=""
              />
            </div>
            <div
              className="mb-3"
              id="password"
            >
              <label
                className="form-label"
              >
                Password
              </label>
              <input
                className="form-control"
                name="newPassword1"
                onChange={[Function]}
                type="password"
                value=""
              />
            </div>
            <div
              className="mb-3"
              id="password-confirm"
            >
              <label
                className="form-label"
              >
                Password Confirmation
              </label>
              <input
                className="form-control"
                name="newPassword2"
                onChange={[Function]}
                type="password"
                value=""
              />
            </div>
            <div
              className="text-center"
            >
              <button
                className="btn btn-primary"
                disabled={false}
                id="submit-button"
                type="submit"
              >
                Sign Up
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</div>
`;
